common/src/main/java/com/terraformersmc/biolith/impl/biome/subbiome/RatioCriterion.java

```diff
@@ -0,0 +1,137 @@
+package com.terraformersmc.biolith.impl.biome.subbiome;
+
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.MapCodec;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+import com.terraformersmc.biolith.api.biome.subbiome.Criterion;
+import com.terraformersmc.biolith.api.biome.subbiome.CriterionType;
+import com.terraformersmc.biolith.impl.biome.BiolithFittestNodes;
+import com.terraformersmc.biolith.impl.biome.DimensionBiomePlacement;
+import net.minecraft.registry.entry.RegistryEntry;
+import net.minecraft.util.StringIdentifiable;
+import net.minecraft.util.dynamic.Range;
+import net.minecraft.util.math.MathHelper;
+import net.minecraft.world.biome.Biome;
+import net.minecraft.world.biome.source.util.MultiNoiseUtil;
+import org.jetbrains.annotations.Nullable;
+import org.joml.Vector2fc;
+
+public record RatioCriterion(RatioTarget target, Range<Float> allowedValues) implements Criterion {
+    public static final MapCodec<RatioCriterion> CODEC = RecordCodecBuilder.mapCodec(instance -> instance.group(
+        RatioTarget.CODEC.fieldOf("target").forGetter(RatioCriterion::target),
+        Codec.FLOAT.optionalFieldOf("min", Float.MIN_VALUE).forGetter(RatioCriterion::min),
+        Codec.FLOAT.optionalFieldOf("max", Float.MAX_VALUE).forGetter(RatioCriterion::max)
+    ).apply(instance, RatioCriterion::new));
+
+    public RatioCriterion(RatioTarget target, float min, float max) {
+        this(target, new Range<>(min, max));
```

Comment:
Note to self:  either refactor to use Vector2fc here or more likely, Range in DimensionBiomePlacement.